File: apps/backend-python/src/housing_platform/bookings/notifications.py
import uuid

from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from housing_platform.db.models import Bookings, Hosts, Notifications, Profiles, Properties


class BookingNotificationService:
    _DATE_FORMAT = "%b %d, %Y"

    def __init__(self, db: Session) -> None:
        self._db = db

    def notify_booking_request(self, booking: Bookings) -> None:
        property_row = self._load_property(booking.property_id)
        if property_row is None or property_row.host is None:
            return

        host_profile_id = property_row.host.profile_id
        guest_name = self._guest_name(booking.customer_id)

        self._create_notification(
            user_id=host_profile_id,
            notification_type="booking_request",
            title="New booking request",
            body=(
                f"{guest_name} requested to stay at {property_row.title} "
                f"({booking.check_in.strftime(self._DATE_FORMAT)} – "
                f"{booking.check_out.strftime(self._DATE_FORMAT)})"
            ),
            metadata={"bookingId": str(booking.id)},
        )

    def notify_booking_confirmed(self, booking: Bookings) -> None:
        property_row = self._load_property(booking.property_id)
        if property_row is None:
            return

        host_name = _host_display_name(property_row.host)
        if not self._profile_exists(booking.customer_id):
            return

        self._create_notification(
            user_id=booking.customer_id,
            notification_type="booking_confirmed",
            title="Booking confirmed",
            body=(
                f"{host_name} confirmed your booking for {property_row.title}, "
                f"{booking.check_in.strftime(self._DATE_FORMAT)} – "
                f"{booking.check_out.strftime(self._DATE_FORMAT)}"
            ),
            metadata={"bookingId": str(booking.id)},
        )

    def notify_booking_rejected(self, booking: Bookings) -> None:
        property_row = self._load_property(booking.property_id)
        if property_row is None or not self._profile_exists(booking.customer_id):
            return

        self._create_notification(
            user_id=booking.customer_id,
            notification_type="booking_rejected",
            title="Booking request declined",
            body=(
                f"Your request to stay at {property_row.title} "
                f"({booking.check_in.strftime(self._DATE_FORMAT)} – "
                f"{booking.check_out.strftime(self._DATE_FORMAT)}) was declined"
            ),
            metadata={"bookingId": str(booking.id)},
        )

    def _load_property(self, property_id: uuid.UUID) -> Properties | None:
        return self._db.scalar(
            select(Properties)
            .options(selectinload(Properties.host).selectinload(Hosts.profile))
            .where(Properties.id == property_id)
        )

    def _guest_name(self, customer_id: uuid.UUID) -> str:
        profile = self._db.scalar(
            select(Profiles).where(
                Profiles.id == customer_id,
                Profiles.deleted_at.is_(None),
            )
        )
        if profile is None:
            return "A guest"
        name = profile.full_name.strip()
        return name if name else "A guest"

    def _profile_exists(self, profile_id: uuid.UUID) -> bool:
        return (
            self._db.scalar(
                select(Profiles.id).where(
                    Profiles.id == profile_id,
                    Profiles.deleted_at.is_(None),
                )
            )
            is not None
        )

    def _create_notification(
        self,
        *,
        user_id: uuid.UUID,
        notification_type: str,
        title: str,
        body: str,
        metadata: dict[str, str],
    ) -> None:
        self._db.add(
            Notifications(
                user_id=user_id,
                type=notification_type,
                title=title,
                body=body,
                metadata_=metadata,
            )
        )


def _host_display_name(host: Hosts | None) -> str:
    if host is None or not host.display_name:
        return "Your host"
    name = host.display_name.strip()
    return name if name else "Your host"

File: apps/backend-go/internal/bookings/notifications.go
package bookings

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgtype"

	"github.com/housing-platform/backend-go/internal/db"
	"github.com/housing-platform/backend-go/internal/shared"
)

const dateFormat = "Jan 02, 2006"

type NotificationService struct{}

func NewNotificationService() *NotificationService {
	return &NotificationService{}
}

func (s *NotificationService) NotifyBookingRequest(ctx context.Context, q pgx.Tx, booking db.Booking) {
	property, hostProfileID := s.loadPropertyHost(ctx, q, booking.PropertyID)
	if property == nil || !hostProfileID.Valid {
		return
	}

	guestName := s.guestName(ctx, q, booking.CustomerID)
	body := fmt.Sprintf("%s requested to stay at %s (%s – %s)",
		guestName, property.Title,
		formatDate(booking.CheckIn), formatDate(booking.CheckOut))

	s.createNotification(ctx, q, hostProfileID, "booking_request", "New booking request", body, booking.ID)
}

func (s *NotificationService) NotifyBookingConfirmed(ctx context.Context, q pgx.Tx, booking db.Booking) {
	property, _ := s.loadPropertyHost(ctx, q, booking.PropertyID)
	if property == nil || !s.profileExists(ctx, q, booking.CustomerID) {
		return
	}

	hostName := s.hostDisplayName(ctx, q, property.HostID)
	body := fmt.Sprintf("%s confirmed your booking for %s, %s – %s",
		hostName, property.Title,
		formatDate(booking.CheckIn), formatDate(booking.CheckOut))

	s.createNotification(ctx, q, booking.CustomerID, "booking_confirmed", "Booking confirmed", body, booking.ID)
}

func (s *NotificationService) NotifyBookingRejected(ctx context.Context, q pgx.Tx, booking db.Booking) {
	property, _ := s.loadPropertyHost(ctx, q, booking.PropertyID)
	if property == nil || !s.profileExists(ctx, q, booking.CustomerID) {
		return
	}

	body := fmt.Sprintf("Your request to stay at %s (%s – %s) was declined",
		property.Title, formatDate(booking.CheckIn), formatDate(booking.CheckOut))

	s.createNotification(ctx, q, booking.CustomerID, "booking_rejected", "Booking request declined", body, booking.ID)
}

type propertyHostInfo struct {
	Title  string
	HostID pgtype.UUID
}

func (s *NotificationService) loadPropertyHost(ctx context.Context, q pgx.Tx, propertyID pgtype.UUID) (*propertyHostInfo, pgtype.UUID) {
	var info propertyHostInfo
	var hostProfileID pgtype.UUID
	err := q.QueryRow(ctx, `
		SELECT p.title, p.host_id, h.profile_id
		FROM public.properties p
		JOIN public.hosts h ON h.id = p.host_id
		WHERE p.id = $1
	`, propertyID).Scan(&info.Title, &info.HostID, &hostProfileID)
	if err != nil {
		return nil, pgtype.UUID{}
	}
	return &info, hostProfileID
}

func (s *NotificationService) guestName(ctx context.Context, q pgx.Tx, customerID pgtype.UUID) string {
	var fullName *string
	err := q.QueryRow(ctx, `
		SELECT full_name FROM public.profiles
		WHERE id = $1 AND deleted_at IS NULL
	`, customerID).Scan(&fullName)
	if err != nil || fullName == nil || *fullName == "" {
		return "A guest"
	}
	return *fullName
}

func (s *NotificationService) hostDisplayName(ctx context.Context, q pgx.Tx, hostID pgtype.UUID) string {
	var displayName *string
	err := q.QueryRow(ctx, `
		SELECT display_name FROM public.hosts WHERE id = $1 AND deleted_at IS NULL
	`, hostID).Scan(&displayName)
	if err != nil || displayName == nil || *displayName == "" {
		return "Your host"
	}
	return *displayName
}

func (s *NotificationService) profileExists(ctx context.Context, q pgx.Tx, profileID pgtype.UUID) bool {
	var exists bool
	_ = q.QueryRow(ctx, `
		SELECT EXISTS(SELECT 1 FROM public.profiles WHERE id = $1 AND deleted_at IS NULL)
	`, profileID).Scan(&exists)
	return exists
}

func (s *NotificationService) createNotification(ctx context.Context, q pgx.Tx, userID pgtype.UUID, notifType, title, body string, bookingID pgtype.UUID) {
	metadata, _ := json.Marshal(map[string]string{
		"bookingId": shared.UUIDToString(bookingID),
	})
	_, _ = q.Exec(ctx, `
		INSERT INTO public.notifications (user_id, type, title, body, metadata)
		VALUES ($1, $2, $3, $4, $5)
	`, userID, notifType, title, body, metadata)
}

func formatDate(d pgtype.Date) string {
	if !d.Valid {
		return ""
	}
	return d.Time.Format(dateFormat)
}
